File: resources/LeptonInjector/resources/examples/inject_muons.py
# Benjamin Smithers
# benjamin.smithers@mavs.uta.edu

# this example script ...
#   + imports the LeptonInjector libraries
#   + creates two injectors, and their operator
#   + tells the operator to execute the process

from math import pi
from pathlib import Path

import EarthModelService as em
import LeptonInjector as LI

# use this if you are on the Cobalt testbed
# xs_folder = "/cvmfs/icecube.opensciencegrid.org/data/neutrino-generator/cross_section_data/csms_differential_v1.0"

# for now, just use the example cross sections that come pre-installed.
# this looks in the parent folder to this file's containing folder
xs_folder = str(Path(__file__).resolve().parent.parent)

# Now, we'll make a new injector for muon tracks
n_events = 55000
diff_xs = xs_folder + "/test_xs.fits"
total_xs = xs_folder + "/test_xs_total.fits"
is_ranged = True
final_1 = LI.Particle.ParticleType.MuMinus
final_2 = LI.Particle.ParticleType.Hadrons
the_injector = LI.Injector(n_events, final_1, final_2, diff_xs, total_xs, is_ranged)


deg = pi / 180.0

# define some defaults
minE = 1000.0  # [GeV]
maxE = 100000.0  # [GeV]
gamma = 2.0
minZenith = 80.0 * deg
maxZenith = 180.0 * deg
minAzimuth = 0.0 * deg
maxAzimuth = 180.0 * deg

# construct the controller
controller = LI.Controller(
    the_injector, minE, maxE, gamma, minAzimuth, maxAzimuth, minZenith, maxZenith
)

# specify the output, earth model
earth_params = str(Path(__file__).resolve().parent.parent / "earthparams") + "/"
earth = em.EarthModelService(
    "earth",
    earth_params,
    ["PREM_mmc"],
    ["Standard"],
    "SimpleIceCap",
    20 * LI.Constants.degrees,
    3500 * LI.Constants.m,
)

controller.SetEarthModelService(earth)
controller.NameOutfile("./data_output.h5")
controller.NameLicFile("./config.lic")

# run the sim
controller.Execute()

File: resources/LeptonInjector/resources/examples/inject_muons.cpp
#include <LeptonInjector/Controller.h>
#include <LeptonInjector/Particle.h>
#include <LeptonInjector/LeptonInjector.h>
#include <LeptonInjector/Constants.h>
#include <string>

/*
This is an example constructed similarly to the python example.

It prepares a run with two injectors, one with CC muon events resulting from a neutrino interaction and one from a nubar interaction.

The output then creates two files:
    - "data_output_c.h5" which has all the data about the interactions
    - "confic_c.lic" which contains a 'snapshot' of the generation settings used to make the aforementioned data. This is to be loaded into LeptonWeighter to weight the events
*/

int main(void){

    // define some parameters shared by both of the injectors 
    int n_events = 25000;
    std::string diff_xs = "../test_xs.fits";
    std::string total_xs = "../test_xs_total.fits";
    bool is_ranged = true;

    // specify the final state particles, and construct the first injector 
    LeptonInjector::Particle::ParticleType final_1 = LeptonInjector::Particle::MuMinus;
    LeptonInjector::Particle::ParticleType final_2 = LeptonInjector::Particle::Hadrons;
    LeptonInjector::Injector the_injector( n_events, final_1, final_2, diff_xs, total_xs, is_ranged);

    // specify more final state particles and contruct the second injector 
    LeptonInjector::Particle::ParticleType final_3 = LeptonInjector::Particle::MuPlus;
    LeptonInjector::Injector next_injector( n_events, final_3, final_2, diff_xs, total_xs, is_ranged);

    // some global values shared by all the injectors 
    // units come from Constants.h 
    double minE = 1000.*LeptonInjector::Constants::GeV;
    double maxE = 100000.*LeptonInjector::Constants::GeV;
    double gamm = 2.;
    double minZenith = 80.*LeptonInjector::Constants::deg;
    double maxZenith = 180.*LeptonInjector::Constants::deg;
    double minAzimuth = 0.*LeptonInjector::Constants::deg;
    double maxAzimuth = 180.*LeptonInjector::Constants::deg;

    // build the Controller object. This will facilitate the simulation itself
    // We need to pass the first injector while building this Controller 
    LeptonInjector::Controller cont(the_injector, minE, maxE, gamm, minAzimuth, maxAzimuth, minZenith, maxZenith, 1200.*LeptonInjector::Constants::m, 1200.*LeptonInjector::Constants::m, 1200.*LeptonInjector::Constants::m, 1200.*LeptonInjector::Constants::m);

    // Add the second injector, specify the output 
    cont.AddInjector(next_injector);
    cont.NameOutfile("./data_output_c.h5");
    cont.NameLicFile("./config_c.lic");

    // Run the program. 
    cont.Execute();
}
